### DISCUSSION TITLE
[Question] How to set windows title back to default after `kitty @ set-window-title xxxxx`

### DISCUSSION THREAD

**Discussion by ray-x** [kovidgoyal/kitty#5925] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/1681295/214163830-4cb7a08b-aaa0-4589-bc36-50bf4ef03074.png)
I set the title of the kitty window inside vim. After exit vim, how do I reset the title back to default (current folder)?

I think one solution is with `kitty @set-window-title` without arguments. But Seems it not work perfectly in split sessions:
After I quit the neovim, and reset the title, right side works fine(1st image) but left side of split still showed the modified title(2nd image).
![image](https://user-images.githubusercontent.com/1681295/214165521-d197790d-1388-4bc2-8178-aa5411170f62.png)
![image](https://user-images.githubusercontent.com/1681295/214165310-a220b973-9c9a-4a37-bf24-7719f84140b1.png)

**Comment by kovidgoyal** [comment]
I dont follow your question. Setting title with escape codes will always work unless you have overridden the title using --title or @ set-window-title. Dont do either of those and it will work. If you do want to use @ set-window-title then use --math=all

**Comment by ray-x** [comment]
In the above example. I run 2 vim instances in kitty horizontal split. Both of these will set the window title to filename/function-name whenever the vim instance is active/on focus by calling `kitty @set-window-title filename->function-name`. 
On vim exit, I will call `kitty @set-window-title` to reset the title back. I did not want to use `--match=all` as there will be other sessions/tabs running vim.
The issue I have is as shown in 2nd screenshot, I run `kitty @set-window-title` in both shell instances/splits. But the title of the window was not fully reset.

**Reply by kovidgoyal** [reply]
I cannot replicate.

**Comment by ray-x** [comment]
It is kind of tricky to reproduce. Please refer to the attached video for details.


https://user-images.githubusercontent.com/1681295/214204910-b295b754-0ae0-4067-b677-3a20cb3c6d8f.mov

**Comment by page-down** [comment, marked as answer]
@ray-x

Did you notice that the titles became very inconsistent?
The window title on the left is set to the file name on the right.

Based on your video, one possibility I can think of is that after resetting the title, the left window was set to the right title again, so it looks like it was not restored.
Especially when the python interpreter starts slowly and a lot of remote control processes are started at the same time.

You need to make sure that there are no more open, unhandled set title processes or commands before resuming the title and before the next shell prompt sends the title.

Overall, it's a matter of timing. For this scenario, you can just use the escape sequence in vim directly without the remote control command. So when closing vim, your title will be set when the next shell prompt appears.

**Reply by ray-x** [reply]
yes, that was something under development.  I want to turn the title to a status line. So when the cursor moves, the window's title will update. As there are multiple vim instances,  you might see the title might not be constant.
But either way, I need to find an option that I can reset the title to default once I exit vim.
I know I am kind of abusive. And if there is only one instance of vim. There is no issue.

**Reply by kovidgoyal** [reply]
I dont follow if you want to set the window title to a statusline, simply emit the escape code to set the window title in vim. There is no need for remote control at all in this scenario. The overall window title will be the title of whichever split has focus.

**Reply by ray-x** [reply]
https://user-images.githubusercontent.com/1681295/214259982-c525d160-d25a-4de8-b52b-b76eee960875.mp4

Please refer to the video. set-windows-title API is closer to what I want.  The escape code seems not working well inside vim.

**Reply by page-down** [reply]
I think this confusion comes from a different understanding of the concept.

Each `split` in kitty is a `kitty window`. And `kitty @ set-window-title` is set for each `kitty window`.
The `OS window` you see will be set to the title of the current `kitty window` when the focus is switched.
https://sw.kovidgoyal.net/kitty/glossary/#term-window

> ... I need to find an option that I can reset the title  ...

If you send the escape sequence in your first screenshot in vim (Although you said `vim`, I think it was `neovim`.), then the problem mentioned in this discussion will not exist.

> Please refer to the video. ... not working well inside vim.

Although I took the time to watch the video, I still don't know what you're doing.
If you want to get help faster, it would be better to use text to describe reproducible steps.

And I'm too unconvinced that vim can't send escape sequences.
Not to mention that in vim you don't have to send escape sequences to set the title yourself at all, just change the variables, assuming you're using a modern version of vim.

**Reply by ray-x** [reply]
Lets start with what I am trying to achieve. 
I want to use the kitty os windows title bar area as a replacement for nvim status line. I want to show dynamic information in kitty title area. Which includes (but is not limited to) file info, the function I am editing, background job results. Those info is dynamic and will be updated using a timer.

So I hook my code into the status line plugin https://github.com/windwp/windline.nvim. When the status line update, The windows title update as well.

In short, the kitty os windows title will be set at 50ms intervals using dynamic text/emoji etc

But once nvim exit. I want to set it back to kitty default.

It works fine if I only have one nvim instance. But if I have multiple nvim instances running, when all nvims exit, the title of the window is not fully reset. 

I know os windows title area is not designed to update at a relatively high frequency by multiple nvim instances. But the real issue I am facing is `kitty @ set-window-title` failed to reset the title back to default if there are other process update the window title at same time.

What I really want is a way to reset windows title back to is original status for the tab calling it.

**Reply by page-down** [reply]
@ray-x

> ... os windows title ... update at a ... high frequency by multiple nvim ...

This is perfectly OK, but I recommend not using timer, but only when the cursor changes position and when the background job status changes.

> ... the real issue ... `kitty @ set-window-title` failed to reset ... if there are other process ...

No, it's not. No offense, this is a problem you created that could have been avoided at the beginning.
The remote control command resets title correctly and sets a new one after that.
You cannot expect this command to reset the title and prohibit updating it for a period of time.

> What I really want is a way to reset windows title back ...
> ...
> It might be related to how fish generate the window title.

No, you don't need to do anything, and it has nothing to do with whether the shell updates the title again.
Because neovim **automatically restores the title** from the "stack" on exit.
Search for `restore title` in neovim's code base or GitHub issues to learn more.

Not to mention that `fish_title` will be called automatically after the command is executed to update the title.

> send escape through ... inside nvim, has no effect

I can't reproduce this.

```log
bash --noprofile --norc
printf "\e]2;title-from-bash\a"

nvim --clean
:set title
:let &titlestring = "hello from nvim titlestring"
:q

nvim --clean
:lua io.write("\27]2;hello-from-nvim\7")
:q
```
After neovim exits, the title is `title-from-bash`.

I hope these explanations answer your questions, if not, then perhaps asking in the neovim/vim project community will get you answers.

**Reply by ray-x** [reply]
Thanks. The scenario you sent is very helpful. I do not even know `titlestring` option before. That might be the best solution I should implement.

**Comment by kovidgoyal** [comment]
For the last time, you are making this unneccessarily complicated.
Simply use the escape codes which you can send in neovim using a
channel. There is no need to involve remote control at all.

**Reply by ray-x** [reply]
Thanks for the suggestions. Apologies for some of the questions as  I do not have much terminal control experience. 
From what I saw
1) send escape code directly inside nvim; not working as nvim hook stdout to nvim console output
2) send escape through lua `io.stdout:write()` inside nvim, has no effect
3) send through job creation and chansend() inside nvim.  has no effect on the title

It might be related to how fish generate the window title. 
The workaround I found is to call `fish_title` when the title changed and needs to rewrite.
